lib/modules/manager/npm/post-update/npm.ts

```diff
@@ -81,14 +85,48 @@ export async function generateLockFile(
 
     // rangeStrategy = update-lockfile
     const lockUpdates = upgrades.filter((upgrade) => upgrade.isLockfileUpdate);
-    if (lockUpdates.length) {
+
+    // divide the deps in two categories: workspace and root
+    const { lockRootUpdates, lockWorkspacesUpdates, workspaces, rootDeps } =
+      divideWorkspaceAndRootDeps(lockFileDir, lockUpdates);
+
+    if (workspaces.size && lockWorkspacesUpdates.length) {
+      logger.debug('Performing lockfileUpdate (npm-workspaces)');
+      for (const workspace of workspaces) {
+        const currentWorkspaceUpdates = lockWorkspacesUpdates
+          .filter((update) => update.workspace === workspace)
+          .filter(
+            (update) =>
+              !rootDeps.has(
+                `${generatePackageKey(update.packageName!, update.newVersion!)}`
```

Comment:
shouldn't you first map to string and then filter?

that would also simplify the join below